### PR TITLE
Expose the mapper to derivers, moving some code out of main.

### PR DESCRIPTION
**Pull request by paurkedal**
Ad issue #33, maybe some final adjustments?

### CODE REVIEW

File: src/ppx_deriving.ml [hunk 1]
```diff
@@ -2,6 +2,7 @@ open Longident
 open Location
 open Asttypes
 open Parsetree
+open Ast_mapper
```
Comment (whitequark) [inline src/ppx_deriving.ml]: You don't have to open the Ast_mapper, all references are qualified

On the other hand you can even remove Ast_helper below

File: src/ppx_deriving.ml [hunk 2]
```diff
@@ -304,3 +305,181 @@ let extract_typename_of_type_group deriver ~allow_shadowing type_list =
...
+  | Some _ -> assert false
+  | None -> []
+
+let add_plugins plugins =
+  let loaded  = get_plugins () in
+  let plugins = List.filter (fun file -> not (List.mem file loaded)) plugins in
+  List.iter dynlink plugins;
+  let loaded  = loaded @ plugins in
+  Ast_mapper.set_cookie "ppx_deriving"
+    (Exp.tuple (List.map (fun file -> Exp.constant (Const_string (file, None))) loaded))
```
Comment (whitequark) [inline src/ppx_deriving.ml]: All Dynlink-related stuff should live in Ppx_deriving_main. Ppx_deriving and the library shouldn't even depend on Dynlink

File: src/ppx_deriving.ml (hunk 1 above)
Comment (paurkedal) [inline src/ppx_deriving.ml]: They are both used multiple places in `ppx_deriving`, esp. I would prefer to keep Ast_mapper due to the record field uses.  OTOH, I could remove all open statements from `ppx_deriving_main`.

File: src/ppx_deriving.ml (hunk 2 above)
Comment (paurkedal) [inline src/ppx_deriving.ml]: Are you sure?  There is a match on @@@findlib.ppxopt in the `structure` local function which calls `add_plugins` and which calls `dynlink`.  I guess it's there to support the `#ppx_deriving` directives in the toplevel.

File: src/ppx_deriving.ml (hunk 1 above)
Comment (whitequark) [inline src/ppx_deriving.ml]: Oh. Maybe just add Ast_mapper.{} around that? I would really prefer not to have get_cookie and such in global scope, it's dirty enough anyway.

File: src/ppx_deriving.ml (hunk 2 above)
Comment (whitequark) [inline src/ppx_deriving.ml]: Oh, yeah, that should also go into `_main`. `Ppx_deriving` itself should really be free of any ocaml/ocamlc-related invocation issues, that would make reusing it much cleaner.

Comment (whitequark) [inline src/ppx_deriving.ml]: Basically, `Ppx_deriving` is an AST mapper, `Ppx_deriving_main` is an AST mapper *specifically for the default OCaml toolchain*.

**Comment by paurkedal** [comment]
I've qualified all Ast_mapper reference and moved dynlink stuff back to main.  In doing the latter, I changed recursion of structure and signature so that they use the mapper.  That allowed me to hook in the ppxopt from main.  I'm not sure how to test the toplevel support though.  I guessed `#ppx "deriving"` but I'm not sure how ppxopt is passed.

**Comment by whitequark** [comment]
Nope. Toplevel is tested in the exact same way as non-toplevel: with findlib. You do `#require "ppx_deriving.std";` for example.

**Comment by paurkedal** [comment]
That explains why `#ppx` is not documented. Digging around a bit, I see that `topfind.ml.in` injects `@@@findlib.ppxopt` to communicate with the syntax extension.

**Comment by whitequark** [comment]
I verified it works on toplevel. Merging, thanks!

**Comment by whitequark** [comment]
It is an overlook that `#ppx` is not documented. It is equivalent to a documented `ocamlc -ppx` option.

**Comment by paurkedal** [comment]
Right, I did test that, as well. And I also have uncomitted code which uses the mapper.

Thanks!

**Pull request merged**